tsdb/head.go

```diff
@@ -680,38 +680,73 @@ func (h *Head) loadWAL(r *wal.Reader, multiRef map[uint64]uint64, mmappedChunks
 		}
 	}()
 
+	// The records are always replayed from the oldest to the newest.
 Outer:
 	for d := range decoded {
 		switch v := d.(type) {
 		case []record.RefSeries:
-			for _, s := range v {
-				series, created, err := h.getOrCreateWithID(s.Ref, s.Labels.Hash(), s.Labels)
+			for _, walSeries := range v {
+				mSeries, created, err := h.getOrCreateWithID(walSeries.Ref, walSeries.Labels.Hash(), walSeries.Labels)
 				if err != nil {
 					seriesCreationErr = err
 					break Outer
 				}
 
+				if h.lastSeriesID.Load() < walSeries.Ref {
+					h.lastSeriesID.Store(walSeries.Ref)
+				}
+
+				mmc := mmappedChunks[walSeries.Ref]
+
 				if created {
-					// If this series gets a duplicate record, we don't restore its mmapped chunks,
-					// and instead restore everything from WAL records.
-					series.mmappedChunks = mmappedChunks[series.ref]
+					// This is the first WAL series record for this series.
+					h.metrics.chunksCreated.Add(float64(len(mmc)))
+					h.metrics.chunks.Add(float64(len(mmc)))
+					mSeries.mmappedChunks = mmc
+					continue
+				}
 
-					h.metrics.chunks.Add(float64(len(series.mmappedChunks)))
-					h.metrics.chunksCreated.Add(float64(len(series.mmappedChunks)))
+				// There's already a different ref for this series.
+				// A duplicate series record is only possible when the old samples were already compacted into a block.
+				// Hence we can discard all the samples and m-mapped chunks replayed till now for this series.
 
-					if len(series.mmappedChunks) > 0 {
-						h.updateMinMaxTime(series.minTime(), series.maxTime())
-					}
-				} else {
-					// TODO(codesome) Discard old samples and mmapped chunks and use mmap chunks for the new series ID.
+				multiRef[walSeries.Ref] = mSeries.ref
 
-					// There's already a different ref for this series.
-					multiRef[s.Ref] = series.ref
+				// It is possible that some old sample is being processed in processWALSamples that could
+				// could cause race below. So we wait for the goroutine to empty input the buffer.
+				idx := mSeries.ref % uint64(n)
+				inputs[idx] <- []record.RefSample{} // To ensure that all the old samples are processed.
+				for len(inputs[idx]) != 0 {
```

Comment:
This is first time I see this pattern of checking len on channel, nice. Isn't that racy, when writer is not fast enough to give elements to input[idx] chan and buffer is read faster than writer writes?

Reply:
The writer is giving elements to `input[idx]` in the same loop, and it is serial to this line. So when we are here, there is nothing giving elements to any channels in `input` slice. So there won't be any read-write race, there is only reading.